app/controllers/coronavirus_form/product_details_controller.rb

```diff
@@ -37,43 +38,45 @@ def destroy
   helper_method :selected_ppe?
 
   def selected_ppe?
-    @product["medical_equipment_type"] == I18n.t(
+    @product[:medical_equipment_type] == I18n.t(
       "coronavirus_form.questions.medical_equipment_type.options.number_ppe.label",
     )
   end
 
   def selected_made_in_uk?
-    @product["product_location"] == I18n.t(
+    @product[:product_location] == I18n.t(
       "coronavirus_form.questions.product_details.product_location.options.option_uk.label",
     )
   end
 
-  def validate_fields(product)
-    missing_fields = validate_missing_fields(product)
-    product_equipment_type = selected_ppe? ? validate_radio_field("#{controller_name}.equipment_type", radio: product["equipment_type"]) : []
-    product_location_validation = validate_radio_field("#{controller_name}.product_location", radio: product["product_location"])
-    postcode_validation = validate_product_postcode(product)
-    missing_fields + product_equipment_type + product_location_validation + postcode_validation
+  def validate_fields
+    [
+      validate_field_response_length(controller_name, TEXT_FIELDS),
+      validate_missing_fields,
+      selected_ppe? ? validate_radio_field("#{controller_name}.equipment_type", radio: @product[:equipment_type]) : [],
+      validate_radio_field("#{controller_name}.product_location", radio: @product[:product_location]),
+      validate_product_postcode,
+    ].compact.flatten
   end
 
-  def validate_product_postcode(product)
+  def validate_product_postcode
     return [] unless selected_made_in_uk?
 
-    if product["product_postcode"].blank?
+    if @product[:product_postcode].blank?
       return [{
-        field: "product_postcode".to_s,
+        field: "product_postcode",
         text: t("coronavirus_form.questions.#{controller_name}.product_location.options.option_uk.input.custom_error"),
       }]
     end
 
-    validate_postcode("product_postcode", product["product_postcode"])
+    validate_postcode("product_postcode", @product[:product_postcode])
   end
 
-  def validate_missing_fields(product)
+  def validate_missing_fields
     required = []
     required.concat REQUIRED_FIELDS
     required.each_with_object([]) do |field, invalid_fields|
-      next if product[field].present?
+      next if @product[field.to_sym].present?
 
       invalid_fields << {
         field: field.to_s,
```

Comment:
Not a blocker: Another sneaky `.to_s`. The Copy and Paste is strong with this one.

Reply:
I'm not sure why `to_s` is used here anyway .... it's already a string 😕